code/modules/awaymissions/zlevel.dm

```diff
@@ -42,38 +66,87 @@
 
 		var/pos = findtext(t, " ")
 		var/name = null
-	//	var/value = null
 
 		if (pos)
 			name = lowertext(copytext(t, 1, pos))
-		//	value = copytext(t, pos + 1)
+
 		else
 			name = lowertext(t)
 
 		if (!name)
 			continue
 
-		potentialRandomZlevels.Add(t)
+		potentialMaps.Add(t)
 
+	return potentialMaps
 
-	if(potentialRandomZlevels.len)
-		var/watch = start_watch()
-		log_startup_progress("Loading away mission...")
 
-		var/map = pick(potentialRandomZlevels)
-		var/file = file(map)
-		if(isfile(file))
-			maploader.load_map(file, do_sleep = 0)
-			late_setup_level(block(locate(1, 1, world.maxz), locate(world.maxx, world.maxy, world.maxz)))
-			log_to_dd("  Away mission loaded: [map]")
+/proc/seedRuins(z_level = 1, budget = 0, whitelist = /area/space, list/potentialRuins = space_ruins_templates)
+	var/overall_sanity = 100
+	var/ruins = potentialRuins.Copy()
+	var/initialbudget = budget
+	var/watch = start_watch()
 
-		for(var/obj/effect/landmark/L in landmarks_list)
-			if (L.name != "awaystart")
+	log_startup_progress("Loading ruins...")
+
+	while(budget > 0 && overall_sanity > 0)
+		// Pick a ruin
+		var/datum/map_template/ruin/ruin = ruins[pick(ruins)]
+		// Can we afford it
+		if(ruin.cost > budget)
+			overall_sanity--
+			continue
+		// If so, try to place it
+		var/sanity = 100
+		// And if we can't fit it anywhere, give up, try again
+
+		while(sanity > 0)
+			sanity--
+			var/turf/T = locate(rand(25, world.maxx - 25), rand(25, world.maxy - 25), z_level)
+			var/valid = 1
+
+			for(var/turf/check in ruin.get_affected_turfs(T,1))
+				var/area/new_area = get_area(check)
+				if(!(istype(new_area, whitelist)))
+					valid = 0
+					break
+
+			if(!valid)
 				continue
-			awaydestinations.Add(L)
 
-		log_startup_progress("  Away mission loaded in [stop_watch(watch)]s.")
+			log_to_dd("  Ruin \"[ruin.name]\" loaded in [stop_watch(watch)]s at ([T.x], [T.y], [T.z]).")
 
-	else
-		log_startup_progress("  No away missions found.")
-		return
\ No newline at end of file
+			var/obj/effect/ruin_loader/R = new /obj/effect/ruin_loader(T)
+			R.Load(ruins,ruin)
+			budget -= ruin.cost
+			if(!ruin.allow_duplicates)
+				ruins -= ruin.name
+			break
+
+	to_chat(world, "<span class='danger'>  Loaded ruins. Or not.</span>") //So the players don't know if we loaded ruins, but we do have a message
```

Comment:
log_startup_progress logs both to chat and to dd, we already have messages that go to dd. I just wanted to have a message after in-chat "Loading ruins..." so the admins know it worked at least.